### PR TITLE
refactor!: add typescript

### PR DESCRIPTION
**Pull request by santimirandarp**
* Passed the code to Typescript, added some workflows, prettier rules, ignores etc.

* It passes the tests and compiles.

* Some parts were refactored, and all the interfaces and types are now explicit. 

* Only the parser is exported from `index.ts` (and some types to enhance the documentation.)
 
Apart from the tests, the demo was tested:

<details>
<summary>
Original code
</summary>

```
detector_maximum_value [ 130.9263458251953 ]
detector_minimum_value [ -0.1758841574192047 ]
actual_run_time_length [ 1860 ]
actual_delay_time [ 0.012000000104308128 ]
actual_sampling_interval [ 0.4000000059604645 ]
ordinate_values [
   -0.07588416337966919,  -0.07525086402893066,    -0.074043869972229,
   -0.07206946611404419,  -0.06943941116333008,  -0.06686896085739136,
   -0.06511062383651733,  -0.06383657455444336,  -0.06227940320968628,
   -0.05998462438583374, -0.057287514209747314, -0.054135918617248535,
  -0.050202012062072754,  -0.04573911428451538,  -0.04174560308456421,
  -0.038720667362213135,  -0.03596395254135132,  -0.03310292959213257,
   -0.02968311309814453, -0.025548040866851807,  -0.02041459083557129,
  -0.014349818229675293, -0.008225440979003906, -0.002250075340270996,
  0.0030621886253356934,  0.008009374141693115,    0.0131949782371521,
    0.01862645149230957,  0.024631619453430176,  0.031165778636932373,
     0.0388026237487793,    0.0469014048576355,  0.054523348808288574,
      0.061817467212677,   0.06961077451705933,   0.07858872413635254,
    0.08841603994369507,    0.0992119312286377,   0.11114776134490967,
    0.12326985597610474,   0.13481080532073975,   0.14563649892807007,
    0.15609711408615112,   0.16608089208602905,    0.1757591962814331,
    0.18655508756637573,     0.199049711227417,   0.21272897720336914,
    0.22683292627334595,    0.2414509654045105,    0.2566799521446228,
     0.2722218632698059,    0.2876892685890198,    0.3031492233276367,
      0.318966805934906,    0.3350377082824707,   0.35137683153152466,
     0.3677159547805786,   0.38423389196395874,    0.4006624221801758,
    0.41688233613967896,   0.43354183435440063,    0.4507526755332947,
    0.46813488006591797,    0.4850253462791443,    0.5019307136535645,
     0.5192682147026062,     0.536799430847168,    0.5547329783439636,
      0.573180615901947,    0.5926191806793213,     0.612504780292511,
     0.6325393915176392,    0.6524622440338135,       0.6718710064888,
     0.6913021206855774,    0.7111355662345886,    0.7317811250686646,
     0.7529333233833313,    0.7742121815681458,    0.7955804467201233,
     0.8167177438735962,    0.8373335003852844,    0.8572563529014587,
     0.8769407868385315,    0.8968636393547058,    0.9174644947052002,
     0.9385645389556885,    0.9598433971405029,    0.9810924530029297,
     1.0021477937698364,    1.0230169296264648,     1.043274998664856,
      1.063227653503418,     1.082993984222412,    1.1029467582702637,
     1.1227428913116455,    1.1423454284667969,    1.1617988348007202,
     1.1810362339019775,
  ... 4551 more items
]
peak_retention_time [
  196.0651397705078,
  332.5663757324219,
  527.5498657226562,
  709.6469116210938,
  734.9354858398438,
  799.1224365234375,
  1030.1668701171875,
  1177.7596435546875
]
peak_start_time [
  186.81199645996094,
  239.21200561523438,
  502.4119873046875,
  668.0120239257812,
  723.64306640625,
  777.2119750976562,
  989.2119750976562,
  1097.2120361328125
]
peak_end_time [
  220.81201171875,
  471.5176696777344,
  572.4786987304688,
  723.64306640625,
  776.9671020507812,
  831.2120361328125,
  1096.9637451171875,
  1354.81201171875
]
peak_width [
  4.974428176879883,
  62.90694808959961,
  11.93286418914795,
  19.319360733032227,
  20.252155303955078,
  15.918797492980957,
  26.703256607055664,
  30.701683044433594
]
peak_area [
  556.7650146484375,
  419.825439453125,
  66.56610107421875,
  294.513671875,
  244.53054809570312,
  72.32331085205078,
  2314.47509765625,
  3948.423095703125
]
peak_area_percent [
  7.0321502685546875,
  5.302552223205566,
  0.8407546877861023,
  3.719817638397217,
  3.0885121822357178,
  0.913470447063446,
  29.232685089111328,
  49.870059967041016
]
peak_height [
  100.07515716552734,
  5.1860527992248535,
  4.82719612121582,
  13.96805477142334,
  10.82530403137207,
  4.233395099639893,
  80.11235809326172,
  117.0067367553711
]
peak_height_percent [
  29.76352310180664,
  1.5423927307128906,
  1.4356645345687866,
  4.154263019561768,
  3.219572067260742,
  1.2590612173080444,
  23.826353073120117,
  34.799171447753906
]
peak_asymmetry [
  1.4555920362472534,
  0.8351489901542664,
  1.7078173160552979,
  1.0537892580032349,
  1.3648090362548828,
  1.131990909576416,
  1.2623794078826904,
  1.2758623361587524
]
baseline_start_time [
  186.81199645996094,
  239.21200561523438,
  502.4119873046875,
  668.0120239257812,
  723.64306640625,
  777.2119750976562,
  989.2119750976562,
  1097.2120361328125
]
baseline_start_value [
  1.9561424255371094,
  0.9857341647148132,
  1.1277347803115845,
  1.3050734996795654,
  1.4332607984542847,
  1.5562325716018677,
  1.5713646411895752,
  2.192728281021118
]
baseline_stop_time [
  220.81201171875,
  471.5176696777344,
  572.4786987304688,
  723.64306640625,
  776.9671020507812,
  831.2120361328125,
  1096.9637451171875,
  1354.81201171875
]
baseline_stop_value [
  1.1907591819763184,
  1.10896897315979,
  1.183473825454712,
  1.4332607984542847,
  1.556132197380066,
  1.466520071029663,
  2.1926772594451904,
  1.658126711845398
]
peak_start_detection_code [
  'B', '', 'B', '', 'B', '',
  'B', '', 'V', '', 'B', '',
  'B', '', 'B', ''
]
peak_stop_detection_code [
  'B', '', 'B', '', 'B', '',
  'V', '', 'B', '', 'B', '',
  'B', '', 'B', ''
]
migration_time [
  196.0651397705078,
  332.5663757324219,
  527.5498657226562,
  709.6469116210938,
  734.9354858398438,
  799.1224365234375,
  1030.1668701171875,
  1177.7596435546875
]
peak_area_square_root [
  23.595869064331055,
  20.489643096923828,
  8.158804893493652,
  17.161399841308594,
  15.637472152709961,
  8.504311561584473,
  48.10899353027344,
  62.83647918701172
]
manually_reintegrated_peaks [
  0, 0, 0, 0,
  0, 0, 0, 0
]
119.02395629882812
-0.07588416337966919
```

</details>

<details>
<summary>
New code
</summary>

```
detector_maximum_value [ 130.9263458251953 ]
detector_minimum_value [ -0.1758841574192047 ]
actual_run_time_length [ 1860 ]
actual_delay_time [ 0.012000000104308128 ]
actual_sampling_interval [ 0.4000000059604645 ]
ordinate_values [
   -0.07588416337966919,  -0.07525086402893066,    -0.074043869972229,
   -0.07206946611404419,  -0.06943941116333008,  -0.06686896085739136,
   -0.06511062383651733,  -0.06383657455444336,  -0.06227940320968628,
   -0.05998462438583374, -0.057287514209747314, -0.054135918617248535,
  -0.050202012062072754,  -0.04573911428451538,  -0.04174560308456421,
  -0.038720667362213135,  -0.03596395254135132,  -0.03310292959213257,
   -0.02968311309814453, -0.025548040866851807,  -0.02041459083557129,
  -0.014349818229675293, -0.008225440979003906, -0.002250075340270996,
  0.0030621886253356934,  0.008009374141693115,    0.0131949782371521,
    0.01862645149230957,  0.024631619453430176,  0.031165778636932373,
     0.0388026237487793,    0.0469014048576355,  0.054523348808288574,
      0.061817467212677,   0.06961077451705933,   0.07858872413635254,
    0.08841603994369507,    0.0992119312286377,   0.11114776134490967,
    0.12326985597610474,   0.13481080532073975,   0.14563649892807007,
    0.15609711408615112,   0.16608089208602905,    0.1757591962814331,
    0.18655508756637573,     0.199049711227417,   0.21272897720336914,
    0.22683292627334595,    0.2414509654045105,    0.2566799521446228,
     0.2722218632698059,    0.2876892685890198,    0.3031492233276367,
      0.318966805934906,    0.3350377082824707,   0.35137683153152466,
     0.3677159547805786,   0.38423389196395874,    0.4006624221801758,
    0.41688233613967896,   0.43354183435440063,    0.4507526755332947,
    0.46813488006591797,    0.4850253462791443,    0.5019307136535645,
     0.5192682147026062,     0.536799430847168,    0.5547329783439636,
      0.573180615901947,    0.5926191806793213,     0.612504780292511,
     0.6325393915176392,    0.6524622440338135,       0.6718710064888,
     0.6913021206855774,    0.7111355662345886,    0.7317811250686646,
     0.7529333233833313,    0.7742121815681458,    0.7955804467201233,
     0.8167177438735962,    0.8373335003852844,    0.8572563529014587,
     0.8769407868385315,    0.8968636393547058,    0.9174644947052002,
     0.9385645389556885,    0.9598433971405029,    0.9810924530029297,
     1.0021477937698364,    1.0230169296264648,     1.043274998664856,
      1.063227653503418,     1.082993984222412,    1.1029467582702637,
     1.1227428913116455,    1.1423454284667969,    1.1617988348007202,
     1.1810362339019775,
  ... 4551 more items
]
peak_retention_time [
  196.0651397705078,
  332.5663757324219,
  527.5498657226562,
  709.6469116210938,
  734.9354858398438,
  799.1224365234375,
  1030.1668701171875,
  1177.7596435546875
]
peak_start_time [
  186.81199645996094,
  239.21200561523438,
  502.4119873046875,
  668.0120239257812,
  723.64306640625,
  777.2119750976562,
  989.2119750976562,
  1097.2120361328125
]
peak_end_time [
  220.81201171875,
  471.5176696777344,
  572.4786987304688,
  723.64306640625,
  776.9671020507812,
  831.2120361328125,
  1096.9637451171875,
  1354.81201171875
]
peak_width [
  4.974428176879883,
  62.90694808959961,
  11.93286418914795,
  19.319360733032227,
  20.252155303955078,
  15.918797492980957,
  26.703256607055664,
  30.701683044433594
]
peak_area [
  556.7650146484375,
  419.825439453125,
  66.56610107421875,
  294.513671875,
  244.53054809570312,
  72.32331085205078,
  2314.47509765625,
  3948.423095703125
]
peak_area_percent [
  7.0321502685546875,
  5.302552223205566,
  0.8407546877861023,
  3.719817638397217,
  3.0885121822357178,
  0.913470447063446,
  29.232685089111328,
  49.870059967041016
]
peak_height [
  100.07515716552734,
  5.1860527992248535,
  4.82719612121582,
  13.96805477142334,
  10.82530403137207,
  4.233395099639893,
  80.11235809326172,
  117.0067367553711
]
peak_height_percent [
  29.76352310180664,
  1.5423927307128906,
  1.4356645345687866,
  4.154263019561768,
  3.219572067260742,
  1.2590612173080444,
  23.826353073120117,
  34.799171447753906
]
peak_asymmetry [
  1.4555920362472534,
  0.8351489901542664,
  1.7078173160552979,
  1.0537892580032349,
  1.3648090362548828,
  1.131990909576416,
  1.2623794078826904,
  1.2758623361587524
]
baseline_start_time [
  186.81199645996094,
  239.21200561523438,
  502.4119873046875,
  668.0120239257812,
  723.64306640625,
  777.2119750976562,
  989.2119750976562,
  1097.2120361328125
]
baseline_start_value [
  1.9561424255371094,
  0.9857341647148132,
  1.1277347803115845,
  1.3050734996795654,
  1.4332607984542847,
  1.5562325716018677,
  1.5713646411895752,
  2.192728281021118
]
baseline_stop_time [
  220.81201171875,
  471.5176696777344,
  572.4786987304688,
  723.64306640625,
  776.9671020507812,
  831.2120361328125,
  1096.9637451171875,
  1354.81201171875
]
baseline_stop_value [
  1.1907591819763184,
  1.10896897315979,
  1.183473825454712,
  1.4332607984542847,
  1.556132197380066,
  1.466520071029663,
  2.1926772594451904,
  1.658126711845398
]
peak_start_detection_code [
  'B', '', 'B', '', 'B', '',
  'B', '', 'V', '', 'B', '',
  'B', '', 'B', ''
]
peak_stop_detection_code [
  'B', '', 'B', '', 'B', '',
  'V', '', 'B', '', 'B', '',
  'B', '', 'B', ''
]
migration_time [
  196.0651397705078,
  332.5663757324219,
  527.5498657226562,
  709.6469116210938,
  734.9354858398438,
  799.1224365234375,
  1030.1668701171875,
  1177.7596435546875
]
peak_area_square_root [
  23.595869064331055,
  20.489643096923828,
  8.158804893493652,
  17.161399841308594,
  15.637472152709961,
  8.504311561584473,
  48.10899353027344,
  62.83647918701172
]
manually_reintegrated_peaks [
  0, 0, 0, 0,
  0, 0, 0, 0
]
119.02395629882812
-0.07588416337966919

```

</details>

There is no difference from `vimdiff`. Could be good to test some more.

### CODE REVIEW

File: .eslintcache [hunk 1]
```diff
@@ -0,0 +1 @@
...o throw\n */\nexport function notNetcdf(statement: boolean, reason: string) {\n  if (statement) {\n    throw new TypeError(`Not a valid NetCDF v3.x file: ${reason}`);\n  }\n}\n\n/**\n * Moves 1, 2, or 3 bytes to next 4-byte boundary\n * @param buffer - Buffer for the file data\n */\nexport function padding(buffer: IOBuffer) {\n  if (buffer.offset % 4 !== 0) {\n    buffer.skip(4 - (buffer.offset % 4));\n  }\n}\n\n/**\n * Reads the name\n * @param buffer - Buffer for the file data\n * @return Name\n */\nexport function readName(buffer: IOBuffer) {\n  // Read name\n  const nameLength = buffer.readUint32();\n  const name = buffer.readChars(nameLength);\n\n  // validate name\n  // TODO\n\n  // Apply padding\n  padding(buffer);\n  return name;\n}\n","/home/sm/clones/cheminfo/netcdfjs/src/parser.ts",[],[],{"ruleId":"71","severity":1,"message":"72","line":49,"column":3,"nodeType":"73","messageId":"74","endLine":49,"endColumn":25},{"ruleId":"71","severity":1,"message":"75","line":35,"column":3,"nodeType":"73","messageId":"74","endLine":35,"endColumn":10},"no-warning-comments","Unexpected 'todo' comment: 'TODO streaming data'.","Line","unexpectedComment","Unexpected 'todo' comment: 'TODO'."]
```
Comment (targos) [inline .eslintcache:1]: This should be in the .gitignore and not committed.

File: .github/workflows/nodejs.yml [hunk 2]
```diff
@@ -3,10 +3,12 @@ name: Node.js CI
 on:
   push:
     branches:
-      - master
+      - main
```
Comment (targos) [inline .github/workflows/nodejs.yml:6]: The default branch is `master` here. We shouldn't make this change before renaming it.

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -3,7 +3,7 @@ name: Release
 on:
   push:
     branches:
-      - master
+      - main
```
Comment (targos) [inline .github/workflows/release.yml:6]: ditto

File: .github/workflows/nodejs.yml (hunk 2 above)
Comment (santimirandarp) [reply .github/workflows/nodejs.yml]: Thanks. Maybe we should update documentation to Typedoc action as well.

**Comment by santimirandarp** [comment]
@targos may be worth to test a few more files to ensure the output is equivalent since there are several users. I am searching for some more now.

File: tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,13 @@
+{
+  "compilerOptions": {
+    "esModuleInterop": true,
+    "moduleResolution": "node",
+    "outDir": "lib",
+    "sourceMap": true,
+    "strict": true,
+    "target": "es2020",
+    "noImplicitAny": true,
+    "noImplicitThis":true
```
Comment (targos) [inline tsconfig.json:10]:
```suggestion
```

these are `true` by default when we have `"strict": true`

Comment (santimirandarp) [reply tsconfig.json]: Thanks, read it in the phone and didn't see I could accept the commit. But it was done below.

**Comment by lpatiny** [comment]
Are you able to test this new package in https://github.com/cheminfo/netcdf-gcms ? There are more test cases

**Comment by santimirandarp** [comment]
Tested this way (can be checked):
* Compiled current project to JS
* Included it in https://github.com/cheminfo/netcdf-gcms
* Run the tests modifying the `index`
```javascript
const { NetCDFReader } = require('./lib/index');
const finniganGCMS = require('./finniganGCMS');
const shimadzuGCMS = require('./shimadzuGCMS');
```
all tests pass.

**Comment by lpatiny** [comment]
@santimirandarp Great thanks for all this work and tests.

Is this really a breaking change ? (your first commit is 'refactor!:'.

I could not find what is really breaking and what should be the adaptations required for the users of this library.

**Comment by santimirandarp** [comment]
You are welcome. I will take another look and let you know today.

**Comment by santimirandarp** [comment]
> I could not find what is really breaking and what should be the adaptations required for the users of this library.

I think the main reason why I put `!` was that some values, like here [in the current `header.js`](https://github.com/cheminfo/netcdfjs/blob/master/src/header.js):
```
  let dimList = dimensionsList(buffer);  // returns object or empty array `[ ]`
  header.recordDimension.id = dimList.recordId;
```
so we are accessing the array and not an object. I believe that `a = { b: undefined }` is not exactly equivalent to `a = { }` for the user. [Is this correct?](https://stackoverflow.com/questions/14967535/delete-a-x-vs-a-x-undefined)

I will list this cases  here to be included in the commit description if you think it will be useful. They are not big differences, just a couple more of the kind cited above @lpatiny.

**Pull request closed without merging**